src/transport/SecurePairingSession.h

```diff
@@ -0,0 +1,166 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file defines the CHIP SPAKE2P Session object that provides
+ *      APIs for constructing spake2p messages and establishing encryption
+ *      keys.
+ *
+ */
+
+#ifndef __SECUREPAIRINGSESSION_H__
+#define __SECUREPAIRINGSESSION_H__
+
+#include <core/ReferenceCounted.h>
+#include <crypto/CHIPCryptoPAL.h>
+#include <system/SystemPacketBuffer.h>
+#include <transport/SecureSession.h>
+
+namespace chip {
+
+extern const char * kSpake2pI2RSessionInfo;
+extern const char * kSpake2pR2ISessionInfo;
+
+using namespace Crypto;
+
+class DLL_EXPORT SecurePairingSessionDelegate : public ReferenceCounted<SecurePairingSessionDelegate>
+{
+public:
+    /**
+     * @brief
+     *   Called when pairing session generates a new message that should be sent to peer.
+     *
+     * @param msgBuf the new message that should be sent to the peer
+     */
+    virtual CHIP_ERROR OnNewMessageForPeer(uint8_t msgType, System::PacketBuffer * msgBuf) { return CHIP_NO_ERROR; }
+
+    /**
+     * @brief
+     *   Called when pairing fails with an error
+     *
+     * @param error error code
+     */
+    virtual void OnPairingError(CHIP_ERROR error) {}
+
+    /**
+     * @brief
+     *   Called when the pairing is complete and the new secure session has been established
+     *
+     */
+    virtual void OnPairingComplete() {}
+
+    virtual ~SecurePairingSessionDelegate() {}
+};
+
+class DLL_EXPORT SecurePairingSession
+{
+public:
+    SecurePairingSession(void);
+    SecurePairingSession(SecurePairingSession &&)      = default;
```

Comment:
The default copy constructors and assignment constructors seem dangerous since we have a reference counted object to the delegate.

Should we do = delete instead?

Reply:
The comment applies to the line below rather than the move operators.